File: leetcode/110-balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  a = _is_balanced root
  a[0]
end


def _is_balanced(root)
  return [true, 0] if root.nil?
  left_balanced, left_h = _is_balanced root.left
  return [false, 0] unless left_balanced
  right_balanced, right_h = _is_balanced root.right
  if right_balanced
    [(left_h-right_h).abs <= 1, [left_h, right_h].max + 1]
  else
    [false, 0]
  end
end

File: leetcode/110-balanced-binary-tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
struct Result {
    bool balanced;
    int depth;
    Result(): balanced(true), depth(0) {}
    Result(bool balanced, int depth = 0): balanced(balanced), depth(depth) {}
};

class Solution {
public:
    bool isBalanced(TreeNode *root) {
        return _isBalanced(root).balanced;
    }
private:
    Result _isBalanced(TreeNode *root) {
        if (root == nullptr) {
            return Result();
        }
        Result left = _isBalanced(root->left);
        if (!left.balanced) {
            return Result(false);
        }
        Result right = _isBalanced(root->right);
        if (!right.balanced) {
            return Result(false);
        }
        if (abs(right.depth - left.depth) > 1) {
            return Result(false);
        }
        return Result(true, 1 + max(left.depth, right.depth));
    }
};
